api/cas-server-core-api-authentication/src/main/java/org/apereo/cas/authentication/Credential.java

```diff
@@ -34,4 +34,21 @@
      * is not readily available or meaningful.
      */
     String getId();
+
+    /**
+     * Returns the unique mark that ids the provider that created the credential.
+     *
+     * @return - the mark
+     */
+    default String getMark() {
```

Comment:
Also, `signature` sounds better to me, if I understand the intention of the method. Or something synonymous to that...

Reply:
I previously had this in MultifactorInstanceCredenential and MultifactorInstanceHandler (names like that), but you said that was confusing in earlier review.  

Essentially we need one additional piece of info in the Credential (interface being most accessible ) so that we can match the mfa provider instance that created the credential to be able to resolve the handler that can authenticate it.

What I really wanted was some low level java feature that I could attach metadata to the object instance that would be an identifier to the provider that created it, like their is with File.

Reply:
Gotcha. Sorry about the back and forth. Too many changes to keep track of :) 

If I am thinking about this right, I think we need a combination of what you had before and what exists now. Adding this capability into the `Credential` doesn't make sense API-wise because it applies to every single authentication handler and would be just extra noise. What might make better sense would be to introduce an in-between interface that is implemented by all MFA credentials (possibly that's what you had?). In that interface, you could add the `mark` method, which really sounds like it should be `multifactorProviderSignature` or `multifactorProviderId`, etc. Then cast down to the interface where needed.

I would also go one step further to say that if you need the `mark` to find the provider to then find the handler, then all such operations should probably go into the in-between interface and then used by the likes of Duo. If we end up supporting multiple instances of other providers, they should all be able to use the same functionality in one common spot. (You could have this sort of logic elsewhere as well , provided it's refactored into a common component accessible to every provider, if and when needed)

Reply:
Refactoring these out into a separate interface is indeed a good call and I am all for it.  I like the idea of making it broader and not so branded to mfa.  CredentialMetadata seemed like a place this could be added to and already exists, or some other standard java interface(s),  but I came up empty.